nvflare/app_common/tie/controller.py

```diff
@@ -0,0 +1,565 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import threading
+import time
+from abc import ABC, abstractmethod
+
+from nvflare.apis.client import Client
+from nvflare.apis.controller_spec import ClientTask, Task
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.impl.controller import Controller
+from nvflare.apis.shareable import ReturnCode, Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.apis.utils.reliable_message import ReliableMessage
+from nvflare.app_common.tie.connector import Connector
+from nvflare.fuel.utils.validation_utils import check_number_range, check_positive_number
+from nvflare.security.logging import secure_format_exception
+
+from .applet import Applet
+from .defs import Constant
+
+
+class _ClientStatus:
+    """
+    Objects of this class keep processing status of each FL client during job execution.
+    """
+
+    def __init__(self):
+        # Set when the client's config reply is received and the reply return code is OK.
+        # If the client failed to reply or the return code is not OK, this value is not set.
+        self.configured_time = None
+
+        # Set when the client's start reply is received and the reply return code is OK.
+        # If the client failed to reply or the return code is not OK, this value is not set.
+        self.started_time = None
+
+        # operation of the last request from this client
+        self.last_op = None
+
+        # time of the last op request from this client
+        self.last_op_time = time.time()
+
+        # whether the app process is finished on this client
+        self.app_done = False
+
+
+class TieController(Controller, ABC):
+    def __init__(
+        self,
+        configure_task_name=Constant.CONFIG_TASK_NAME,
+        configure_task_timeout=Constant.CONFIG_TASK_TIMEOUT,
+        start_task_name=Constant.START_TASK_NAME,
+        start_task_timeout=Constant.START_TASK_TIMEOUT,
+        job_status_check_interval: float = Constant.JOB_STATUS_CHECK_INTERVAL,
+        max_client_op_interval: float = Constant.MAX_CLIENT_OP_INTERVAL,
+        progress_timeout: float = Constant.WORKFLOW_PROGRESS_TIMEOUT,
+    ):
+        """
+        Constructor
+
+        Args:
+            configure_task_name - name of the config task
+            configure_task_timeout - time to wait for clients’ responses to the config task before timeout.
+            start_task_name - name of the start task
+            start_task_timeout - time to wait for clients’ responses to the start task before timeout.
+            job_status_check_interval - how often to check client statuses of the job
+            max_client_op_interval - max amount of time allowed between XGB ops from a client
```

Comment:
XGB should be changed

Reply:
Good catch! Changed to replace XGB with "app".